datafusion/functions/src/core/coalesce.rs

```diff
@@ -60,12 +60,16 @@ impl ScalarUDFImpl for CoalesceFunc {
     }
 
     fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
-        Ok(arg_types[0].clone())
+        Ok(arg_types
+            .iter()
+            .find_or_first(|d| !d.is_null())
+            .unwrap()
+            .clone())
     }
 
-    // If all the element in coalesce is non-null, the result is non-null
+    // If all the elements in coalesce are non-null, the result is non-null
     fn is_nullable(&self, args: &[Expr], schema: &dyn ExprSchema) -> bool {
-        args.iter().any(|e| e.nullable(schema).ok().unwrap_or(true))
+        args.iter().all(|e| e.nullable(schema).ok().unwrap_or(true))
```

Comment:
I think the logic here is incorrect.

`e.nullable(schema)` is true if `nullable`, false if `non-null`.

Therefore, whenever we found at least one is `nullable`, we should returns `true`.

Reply:
But wouldn't a coalesce be `non-null` (`is_nullable` -> false) if at least one of the args is `non-null`?

Reply:
You are right, the comment should also be updated.

Reply:
Also, we should fix here too.

We should provide the coerced types to `return_type`

https://github.com/apache/datafusion/blob/f72622d6fcd27adac0bce5e87cccc83bad215448/datafusion/expr/src/expr_schema.rs#L153-L167

Reply:
Great! Where can I find an example of coercion of types?

Reply:
`coalesce(null, 5)` will coerced to i64. Therefore we have i64(None) and i64(5). Is this what you are asking for?

Reply:
Sorry, I was asking more about which function to use to coerce the types.
I see that in the `coerce_types` function, there is the following code:
```rust
if arg_types.is_empty() {
            return exec_err!("coalesce must have at least one argument");
        }
        let new_type = type_union_resolution(arg_types)
            .unwrap_or(arg_types.first().unwrap().clone());
``` 

Should I follow the same pattern?

Reply:
Sorry, but I'm not following. Are you suggesting that we change coerce `arg_data_types` and pass that result to:

```rust
 // perform additional function arguments validation (due to limited 
 // expressiveness of `TypeSignature`), then infer return type 
 Ok(func.return_type_from_exprs(args, schema, &arg_data_types)?) 
```

Reply:
Great! Thanks for the quick feedback!

Reply:
I pushed that change to the PR, but it caused some test failures, for example the statement:
```sql
SELECT encode(12, 'hex')
```

Doesn't produce the error:

> query error DataFusion error: Error during planning: The encode function can only accept utf8 or binary\.

The source seems to be this line in the function `coerced_from`:
https://github.com/apache/datafusion/blob/91b1d2bfe8f603df94e846b91d8475a0af2e5240/datafusion/expr/src/type_coercion/functions.rs#L681-L682

It specifies that any `DataType` can be coerced into a `Utf8`.

So, what is the best approach to move forward, @jayzhan211?

Reply:
We need to fix each function's signature

Reply:
https://github.com/mesejo/arrow-datafusion/pull/1/files

I try to fix the signature of encode to make sure the suggestion makes sense.
I think we might need to fix other functions as well

Reply:
I am having some sqllogictests fail in the CI ([link](https://github.com/mesejo/arrow-datafusion/actions/runs/10728724019/job/29753866598)), but executing:
```bash
cargo test --test sqllogictests
```  
Passes on my local machine. Have you experienced something similar? 

**Note:** The error message says that it only received a prefix of the actual message

Reply:
@mesejo  Try `cargo test --test sqllogictests` (slt files), `cargo test --lib --tests --bins --features avro,json` (rust test) and `./dev/rust_lint.sh` (clippy + fmt)

Reply:
@jayzhan211 Thanks for the tips! I updated the functions; this is ready for review.